sdk/identity/azure-identity/src/workload_identity_credential.cpp

```diff
@@ -6,65 +6,110 @@
-          std::string(
-              "grant_type=client_credentials"
-              "&client_assertion_type="
-              "urn%3Aietf%3Aparams%3Aoauth%3Aclient-assertion-type%3Ajwt-bearer" // cspell:disable-line
-              "&client_id=")
-          + Url::Encode(clientId)),
-      m_tokenFilePath(tokenFilePath)
+      m_clientCredentialCore(tenantIdOptions, authorityHostOptions, additionallyAllowedTenants)
 {
+  std::string tenantId;
+  std::string clientId;
+  std::string authorityHost;
+
+  if (tenantIdOptions.empty())
+  {
+    tenantId = Environment::GetVariable(AzureTenantIdEnvVarName);
+    if (tenantId.empty())
+    {
+      throw std::runtime_error(
+          "No tenant ID specified. Check pod configuration or set TenantID in the options.");
+    }
+  }
+  if (clientIdOptions.empty())
+  {
+    clientId = Environment::GetVariable(AzureClientIdEnvVarName);
+    if (clientId.empty())
+    {
+      throw std::runtime_error(
+          "No client ID specified. Check pod configuration or set ClientID in the options.");
+    }
+  }
+  if (authorityHostOptions.empty())
+  {
+    authorityHost = Environment::GetVariable(AzureAuthorityHostEnvVarName);
+    if (authorityHost.empty())
+    {
+      authorityHost = _detail::ClientCredentialCore::AadGlobalAuthority;
+    }
+  }
+  if (tokenFilePathOptions.empty())
+  {
+    m_tokenFilePath = Environment::GetVariable(AzureFederatedTokenFileEnvVarName);
+    if (m_tokenFilePath.empty())
+    {
+      throw std::runtime_error(
+          "No token file specified. Check pod configuration or set TokenFilePath in the options.");
+    }
+  }
+
+  m_clientCredentialCore = Azure::Identity::_detail::ClientCredentialCore(
+      tenantId, authorityHost, additionallyAllowedTenants);
+  m_tokenCredentialImpl = std::make_unique<TokenCredentialImpl>(options);
+  m_requestBody
+      = std::string(
+            "grant_type=client_credentials"
+            "&client_assertion_type="
+            "urn%3Aietf%3Aparams%3Aoauth%3Aclient-assertion-type%3Ajwt-bearer" // cspell:disable-line
+            "&client_id=")
+      + Url::Encode(clientId);
 }
 
 WorkloadIdentityCredential::WorkloadIdentityCredential(
-    std::string tenantId,
-    std::string const& clientId,
-    std::string const& tokenFilePath,
     WorkloadIdentityCredentialOptions const& options)
     : WorkloadIdentityCredential(
-        tenantId,
-        clientId,
-        tokenFilePath,
+        options.TenantId,
+        options.ClientId,
         options.AuthorityHost,
+        options.TokenFilePath,
         options.AdditionallyAllowedTenants,
         options)
 {
 }
 
 WorkloadIdentityCredential::WorkloadIdentityCredential(
-    std::string tenantId,
-    std::string const& clientId,
-    std::string const& tokenFilePath,
     Core::Credentials::TokenCredentialOptions const& options)
     : WorkloadIdentityCredential(
-        tenantId,
-        clientId,
-        tokenFilePath,
+        WorkloadIdentityCredentialOptions{}.TenantId,
```

Comment:
You are constructing 5 new `WorkloadIdentityCredentialOptions` objects (each of which has 5 strings that need to be constructed and destroyed), just to retrieve the default values for each of the fields.

That seems REALLY complicated and a LOT of code just to retrieve a bunch of empty strings.

Reply:
I agree 100% about this and will fix here.

We should also revisit this coding pattern and fix it across all the credentials.